File: src/python/halchemy/lib/status_codes.py
def parse_status_code_setting(status_code_setting: str):
    """
    Parses a status code setting string into a list of conditions.

    Returns a list of tuples in the form:
    - ('range', start, end) for range conditions
    - ('gt', value, None) for greater-than conditions
    - ('lt', value, None) for less-than conditions
    - ('gte', value, None) for greater-than-or-equal conditions
    - ('lte', value, None) for less-than-or-equal conditions
    - ('eq', value, None) for exact match conditions
    """
    parts = status_code_setting.replace(',', ' ').split()
    ranges = []
    for part in parts:
        part = part.strip()
        if '-' in part:
            start, end = map(int, part.split('-'))
            ranges.append(('range', start, end))
        elif part.startswith('>='):
            value = int(part[2:])
            ranges.append(('gte', value))
        elif part.startswith('<='):
            value = int(part[2:])
            ranges.append(('lte', value))
        elif part.startswith('>'):
            value = int(part[1:])
            ranges.append(('gt', value))
        elif part.startswith('<'):
            value = int(part[1:])
            ranges.append(('lt', value))
        else:
            value = int(part)
            ranges.append(('eq', value))
    return ranges


def do_settings_include_status_code(settings: str, status_code: int):
    """
    Determines if a status code is included in the settings.

    :param settings: The status code settings string (e.g., '>=400', '400-403, >404').
    :param status_code: The status code to check.
    :return: True if the status code matches the settings, False otherwise.
    """
    if settings is None:
        return False

    included = False
    range_setting = parse_status_code_setting(settings)
    for condition, *values in range_setting:
        if condition == 'range' and values[0] <= status_code <= values[1]:
            included = True
        elif condition == 'gt' and status_code > values[0]:
            included = True
        elif condition == 'lt' and status_code < values[0]:
            included = True
        elif condition == 'gte' and status_code >= values[0]:
            included = True
        elif condition == 'lte' and status_code <= values[0]:
            included = True
        elif condition == 'eq' and status_code == values[0]:
            included = True

    return included

File: src/node/halchemy/src/lib/status_codes.ts
type RangePart = ['range', number, number] | ['gt', number] | ['lt', number] | ['gte', number] | ['lte', number] | ['eq', number]

function parseRangeSetting(settings: string): RangePart[] {
    const parts = settings.replace(',', ' ').split(/\s+/)
    const ranges: RangePart[] = []

    parts.forEach(part => {
        if (part.includes('-')) {
            const [start, end] = part.split('-').map(Number)
            ranges.push(['range', start, end])
        } else if (part.startsWith('>=')) {
            const value = parseInt(part.slice(2), 10)
            ranges.push(['gte', value])
        } else if (part.startsWith('<=')) {
            const value = parseInt(part.slice(2), 10)
            ranges.push(['lte', value])
        } else if (part.startsWith('>')) {
            const value = parseInt(part.slice(1), 10)
            ranges.push(['gt', value])
        } else if (part.startsWith('<')) {
            const value = parseInt(part.slice(1), 10)
            ranges.push(['lt', value])
        } else {
            const value = parseInt(part, 10)
            ranges.push(['eq', value])
        }
    })

    return ranges
}


export function doSettingsIncludeStatusCode(settings: string | null, statusCode: number): boolean {
    if (settings === '' || settings === null) {
        return false
    }

    const rangeSetting = parseRangeSetting(settings as string)

    return rangeSetting.some(([condition, ...values]) => {
        switch (condition) {
            case 'range':
                const value = values[1] === undefined? 100000 : values[1]  // TODO: is there a better way to guard against values[1] being undefined?
                return statusCode >= values[0] && statusCode <= value
            case 'gt':
                return statusCode > values[0]
            case 'lt':
                return statusCode < values[0]
            case 'gte':
                return statusCode >= values[0]
            case 'lte':
                return statusCode <= values[0]
            case 'eq':
                return statusCode === values[0]
            default:
                return false
        }
    })
}
